src/main/java/com/squareup/kotlinpoet/TypeSpec.kt

```diff
@@ -431,15 +437,34 @@ class TypeSpec private constructor(builder: TypeSpec.Builder) {
       this.superinterfaces += superinterfaces
     }
 
-    fun addSuperinterface(superinterface: TypeName) = apply {
-      superinterfaces += superinterface
+    fun addSuperinterface(superinterface: TypeName, delegate: CodeBlock = CodeBlock.EMPTY) = apply {
+      if (delegate.isEmpty())
+        superinterfaces += superinterface
+      else {
+        require(kind == Kind.CLASS) {
+          "delegation only allowed for classes (found $kind '$name')" }
+        require(!superinterface.nullable) {
+          "expected non-nullable type but was '${superinterface.asNonNullable()}'"}
+        require(delegatedSuperInterfaces[superinterface] == null) {
+          "'$name' can not delegate to $superinterface by $delegate with existing declaration by " +
+              "${delegatedSuperInterfaces[superinterface]}" }
+        delegatedSuperInterfaces.put(superinterface, delegate)
+      }
     }
 
-    fun addSuperinterface(superinterface: Type)
-        = addSuperinterface(superinterface.asTypeName())
+    fun addSuperinterface(superinterface: Type, delegate: CodeBlock = CodeBlock.EMPTY)
+        = addSuperinterface(superinterface.asTypeName(), delegate)
+
+    fun addSuperinterface(superinterface: KClass<*>, delegate: CodeBlock = CodeBlock.EMPTY)
+        = addSuperinterface(superinterface.asTypeName(), delegate)
 
-    fun addSuperinterface(superinterface: KClass<*>)
-        = addSuperinterface(superinterface.asTypeName())
+    fun addSuperinterface(constructorParameterName: String) = apply {
```

Comment:
I'm still not sure about this behavior. I was thinking that we'd do `KClass` + `String` since most often you won't need to power of a `CodeBlock` to declare a delegate.